tree/ntuple/v7/inc/ROOT/RColumnElement.hxx

```diff
@@ -55,65 +75,126 @@ static void CopyElementsBswap(void *destination, const void *source, std::size_t
    }
 }
 
-/// \brief Re-arranges the bytes of the array of N-byte elements in columnar layout.
+/// Casts T to one of the ints used in RByteSwap and back to its original type, which may be float or double
+#if R__LITTLE_ENDIAN == 0
+template <typename T>
+void ByteSwapIfNecessary(T &value)
+{
+   constexpr auto N = sizeof(T);
+   using bswap_value_type = typename RByteSwap<N>::value_type;
+   void *valuePtr = &value;
+   auto swapped = RByteSwap<N>::bswap(*reinterpret_cast<bswap_value_type *>(valuePtr));
+   *reinterpret_cast<bswap_value_type *>(valuePtr) = swapped;
+}
+#else
+#define ByteSwapIfNecessary(x) ((void)0)
+#endif
+
+/// \brief Pack `count` elements into narrower (or wider) type
+///
+/// Used to convert in-memory elements to smaller column types of comatible types
+/// (e.g., double to float, int64 to int32). Takes care of byte swap if necessary.
+template <typename DestT, typename SourceT>
+static void CastPack(void *destination, const void *source, std::size_t count)
+{
+   auto dst = reinterpret_cast<DestT *>(destination);
+   auto src = reinterpret_cast<const SourceT *>(source);
+   for (std::size_t i = 0; i < count; ++i) {
+      dst[i] = src[i];
+      ByteSwapIfNecessary(dst[i]);
+   }
+}
+
+/// \brief Unpack `count` on-disk elements into wider (or narrower) in-memory array
 ///
-/// Assumes that the elements are stored in little-endian layout. The destination stores
-/// all the first bytes first, then all the second bytes, etc.
-template <std::size_t N>
-static void SplitElementsLE(void *destination, const void *source, std::size_t count)
+/// Used to convert on-disk elements to larger C++ types of comatible types
+/// (e.g., float to double, int32 to int64). Takes care of byte swap if necessary.
+template <typename DestT, typename SourceT>
+static void CastUnpack(void *destination, const void *source, std::size_t count)
 {
-   const char *unsplitArray = reinterpret_cast<const char *>(source);
-   char *splitArray = reinterpret_cast<char *>(destination);
+   auto dst = reinterpret_cast<DestT *>(destination);
+   auto src = reinterpret_cast<const SourceT *>(source);
    for (std::size_t i = 0; i < count; ++i) {
+      DestT val = src[i];
+      ByteSwapIfNecessary(val);
+      dst[i] = val;
```

Comment:
I don't think it would be the same.  E.g., when going back from on-disk float to double, first assigning to `dst[i]` would scramble the value on a byteswapped `val`.

Reply:
Good point! Thanks for spotting! Actually, the corresponding unit test was mixed-up, too. Should be fixed now.